File: AdvertisingPlatforms/AdvertisingPlatforms/wwwroot/js/models/appParameters.js
// Прототип паарметров приложения со значениями по умолчанию
class AppParameters{

    // Список допустимых расширений файла<br/>
    // Если список пуст, то допустимы все расширения
    allowedExtensions = [".txt"];

    // Список допустимых MIME типов файла<br/>
    // Если список пуст, то допустимы все MIME типы
    allowedMimeTypes = ["text/plain"];

    // Максимальный размер файла
    maxSizeFile = 50*1024*124;// 50 Мб

    // Если параметр равен true и CapitaLetterSensitivity = false,
    // то следующие локации одинаковые: /ru /Ru /rU /RU
    // Если параметр равен false и CapitaLetterSensitivity = false
    // то файл с локациями в верхнем регистром не валиден: /Ru /rU /RU -> ошибки
    allowingTheUseOfCapitalLetters = false;

    // Параметр чувствительности к символам верхнего регистра
    // Если параметр равен true, то следующие локации разные: /ru /Ru /rU /RU
    capitaLetterSensitivity = false;

    // Разрешение на использование подлокаций с одинаковыми названиями
    // Если параметр true,
    // то могут существовать одновременно локации типа: /ru/fp, /kz/.../fp -> валидны
    // Иначе говоря, нельзя однозначно определить локацию по последней подлокации
    // /fp -> /ru/fp и /kz/.../fp
    locationsWithTheSameName = false;

    // Разрешение на использование одинаковых подлокаций в полной локации
    // Если параметр равен true и LocationsWithTheSameName = true
    // то следующие локации типа: /ru/ru, /ru/.../ru -> валидны
    // Если параметр равен false и LocationsWithTheSameName = true
    // то следующие локации типа: /ru/ru, /ru/.../ru -> ошибки
    repeatingSubLocations = false;

    // Патерн для валидации разрешённых символов
    // Если AllowingTheUseOfCapitalLetters = true,
    // то значение: /^[a-zA-Z/]+$/ - верхний регистр разрешён
    // иначе значение: /^[a-z/]+$/ - верхний регистр генирирует отслеживаемые ошибки
    stringValidationPattern = /^[a-z/]+$/;

    // Шаблон ответа для описания ошибки данных в файле
    dataTemplateDescription;

    // Шаблон ответа для описания ошибки локации
    locationTemplateDescription;
    
    static GetStringValidationPattern(allowingTheUseOfCapitalLetters){
        return allowingTheUseOfCapitalLetters ? /^[a-zA-Z/]+$/
                                              : /^[a-z/]+$/;
    }
    
    // Метод инициализации шаблонов ответов для ошибок валидации на клиенте
    SetTemplate() {
        var RepeatingSubLocationsDescription = this.locationsWithTheSameName && this.repeatingSubLocations
                    ? "также разрешено использовать повторения в полной локации: /ru/ru или /ru.../gz/ru - допустимо."
                    : "но повторения в полной локации запрещены, например: /ru/ru или /ru.../gz/ru - запрещены.";

        var LocationsWithTheSameNameDescription = this.locationsWithTheSameName
                    ? `- Конечные подлокации могут повторяться, например: /ru и /pz/ru - допустимо,
                       <br>&nbsp;&nbsp;&nbsp;${RepeatingSubLocationsDescription}`
                    : "- Повторяющиеся локации запрещены, например: /ru и /pz/ru , /ru/ru или /ru.../gz/ru - ошибки.";

        this.locationTemplateDescription =
        `Требования к локациям:
        <br>&nbsp;&nbsp;&nbsp;- Локации должны начинаться и отделяться символом '/';
        <br>&nbsp;&nbsp;&nbsp;- Локации задаются символами нижнего ${(this.allowingTheUseOfCapitalLetters ? "и вержнего " : "")}регистра латинского алфавита;
        <br>&nbsp;&nbsp;&nbsp;${LocationsWithTheSameNameDescription}`;

        this.dataTemplateDescription =
        `Рекламные площадки должны быть представлены в виде шаблона:
        <br>&nbsp;&nbsp;&nbsp;<Название площдадки> : <Локации работы площадки>
        <br>Локации работы площадки перечисляются через запятую и задаются следующим шаблоном:
        <br>&nbsp;&nbsp;&nbsp;/<Локация>/<Локация>.../<Локация>
        <br>${this.locationTemplateDescription}`;

        this.stringValidationPattern = this.allowingTheUseOfCapitalLetters ? /^[a-zA-Z/]+$/
                                                                      : /^[a-z/]+$/;
    }
}

File: AdvertisingPlatforms/AdvertisingPlatforms/ApplicationSettings/AppParameters.cs

using AdvertisingPlatforms.Core.Abstractions;

namespace AdvertisingPlatforms.ApplicationSettings
{
    /// <summary>
    /// Реализация интерфейса <see cref="IAppParameters"/> <br/>
    /// Пользовательские настройки приложения
    /// </summary>
    public class AppParametersSingleton : IAppParameters
    {
        public string[] AllowedExtensions { get;  set; }

        public string[] AllowedMimeTypes { get; set; }

        public int MaxSizeFile { get;  set; }

        public bool AllowingTheUseOfCapitalLetters { get;  set; }

        public bool CapitaLetterSensitivity { get;  set; }

        public bool LocationsWithTheSameName { get;  set; }

        public bool RepeatingSubLocations { get;  set; }

        public string StringValidationPattern { get;  set; }

        public string DataTemplateDescription { get;  set; }

        public string LocationTemplateDescription { get; set; }

        #region Реализация паттерна одиночка (Singleton)

        // Свойтво гарантирует патокобезопасность и отложенное создание экземпляра класса
        public static AppParametersSingleton GetInstance { get; } = new();

        #endregion

        /// <summary>
        /// Инициализация пользовательских настроек приложения
        /// </summary>
        public AppParametersSingleton()
        {
            // Получение пользовательских настроек
            IConfiguration configuration = new ConfigurationBuilder()
                .SetBasePath(Directory.GetCurrentDirectory())
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .Build();


            #region Получаем настройки валидации файла

            var fileValidationParameters = configuration.GetSection("ApplicationSettings:FileValidationParameters");


            // Получаем список допустимых расширений файла
            AllowedExtensions = fileValidationParameters.GetSection("AllowedExtensions").Get<string[]>()
                                ?? [".txt"];// По умолчанию


            // Инициализируем список разрешённых MIME типов файлов (разрешённый контекст файла)
            AllowedMimeTypes = fileValidationParameters.GetSection("AllowedMimeTypes").Get<string[]>()
                               ?? ["text/plain"];// По умолчанию


            // Инициализируем ограничение по весу файла
            MaxSizeFile = fileValidationParameters.GetValue<int>("MaxSize");
            if (MaxSizeFile == 0)
            {
                MaxSizeFile = 50 * 1024 * 1024;// 50 Мб - по умолчанию
            }

            #endregion


            #region Получаем настройки валидации рекламных платформ

            var validationParameters = configuration.GetSection("ApplicationSettings:AdvertisingPlatformValidationParameters");

            // Получение параметра на чувствительнось к символам верхнего регистра.
            CapitaLetterSensitivity = validationParameters.GetValue<bool>("CapitaLetterSensitivity");


            // Получение разрешения на использование верхнего регистра
            if (CapitaLetterSensitivity)
            {
                // Если локации чувствительны к верхнему регистру, то верхний регистр автоматически разрешён
                AllowingTheUseOfCapitalLetters = true;
            }
            else
            {
                // Получение параметра на разрешение использования верхнего регистра
                AllowingTheUseOfCapitalLetters = validationParameters.GetValue<bool>("AllowingTheUseOfCapitalLetters");
            }

            // Получение параметра на наличие локаций с одинаковыми названиями
            LocationsWithTheSameName = validationParameters.GetValue<bool>("LocationsWithTheSameName");

            // Получение разрешения на использование одинаковых подлокаций в полной локации
            if (LocationsWithTheSameName)
            {
                // Получение параметра на возможность использовать одинаковые подлокации в полной локации
                RepeatingSubLocations = validationParameters.GetValue<bool>("RepeatingSubLocations");
            }
            else
            {
                // Если запрещено использовать локации с одинаковыми названиями, то повторений быть не может
                RepeatingSubLocations = false;
            }

            // Инициализация патерна для валидации разрешённых символов
            StringValidationPattern = AllowingTheUseOfCapitalLetters
                       ? @"^[a-zA-Z/]+$" // Тут верхний регист разрешён
                       : @"^[a-z/]+$";   // Тут верхний регист вызывает ошибки

            #endregion


            #region Генерируем описание шаблона данных

            string RepeatingSubLocationsDescription = LocationsWithTheSameName && RepeatingSubLocations
                    ? "также разрешено использовать повторения в полной локации:  /ru/ru или /ru.../gz/ru - допустимо."
                    : "но повторения в полной локации запрещены, например: /ru/ru или /ru.../gz/ru - запрещены.";

            string LocationsWithTheSameNameDescription = LocationsWithTheSameName
                    ? $"- Конечные подлокации могут повторяться, например: /ru и /pz/ru - допустимо,"+
                      $"\n\t{RepeatingSubLocationsDescription}"
                    : "- Повторяющиеся локации запрещены, например: /ru и /pz/ru , /ru/ru или /ru.../gz/ru - ошибки.";


            LocationTemplateDescription =
                "Требования к локациям:"+
                "\n\t- Локации должны начинаться и отделяться символом '/';"+
                $"\n\t- Локации задаются символами нижнего {(AllowingTheUseOfCapitalLetters ? "и вержнего " : "")}регистра латинского алфавита;"+
                $"\n\t{LocationsWithTheSameNameDescription}";

            DataTemplateDescription =
                "Рекламные площадки должны быть представлены в виде шаблона:"+
                "\n\t<Название площдадки> : <Локации работы площадки>"+
                "\nЛокации работы площадки перечисляются через запятую и задаются следующим шаблоном:"+
                "\n\t/<Локация>/<Локация>.../<Локация>"+
                $"\n{LocationTemplateDescription}";


            #endregion
        }
    }
}
